drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,342 @@
+#![no_std]
+#![no_main]
+
+use drv_update_api::UpdateError;
+use idol_runtime::{ClientError, Leased, LenLimit, RequestError, R};
+use ringbuf::*;
+use stm32h7::stm32h753 as device;
+use userlib::*;
+
+// Keys constants are defined in RM0433 Rev 7
+// Section 4.9.2
+const FLASH_KEY1: u32 = 0x4567_0123;
+const FLASH_KEY2: u32 = 0xCDEF_89AB;
+
+// Keys constants are defined in RM0433 Rev 7
+// Section 4.9.3
+const FLASH_OPT_KEY1: u32 = 0x0819_2A3B;
+const FLASH_OPT_KEY2: u32 = 0x4C5D_6E7F;
+
+const BANK_ADDR: u32 = 0x08100000;
+const BANK_END: u32 = 0x08200000;
+
+// Writes are indexed by flash words, BANK_ADDR is word 0,
+// BANK_ADDR + FLASH_WORD_BYTES is word 1 etc.
+const BANK_WORD_LIMIT: usize =
+    (BANK_END - BANK_ADDR) as usize / FLASH_WORD_BYTES;
+
+// RM0433 Rev 7 section 4.3.9
+// Flash word is defined as 256 bits
+const FLASH_WORD_BITS: usize = 256;
+
+// Total length of a word in bytes (i.e. our array size)
+const FLASH_WORD_BYTES: usize = FLASH_WORD_BITS / 8;
+
+// number of 32-bit words we will have to write
+const FLASH_CPU_WORDS: usize = FLASH_WORD_BYTES / 4;
+
+// Block is an abstract concept here. It represents the size of data the
+// driver will process at a time.
+const BLOCK_SIZE_BYTES: usize = FLASH_WORD_BYTES * 32;
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    EraseStart,
+    EraseEnd,
+    WriteStart,
+    WriteEnd,
+    FinishStart,
+    FinishEnd,
+    WriteBlock(usize),
+    None,
+}
+
+enum UpdateState {
+    NoUpdate,
+    InProgress,
+    Finished,
+}
+
+ringbuf!(Trace, 64, Trace::None);
+
+const FLASH_TIMEOUT: usize = 10000;
+
+struct ServerImpl<'a> {
+    flash: &'a device::flash::RegisterBlock,
+    state: UpdateState,
+}
+
+impl<'a> ServerImpl<'a> {
+    // See RM0433 Rev 7 section 4.3.13
+    fn swap_banks(&mut self) -> Result<(), RequestError<UpdateError>> {
+        let flash = unsafe { &*device::FLASH::ptr() };
+
+        ringbuf_entry!(Trace::FinishStart);
+        if flash.optsr_cur().read().swap_bank_opt().bit() {
+            flash
+                .optsr_prg()
+                .modify(|_, w| w.swap_bank_opt().clear_bit());
+        } else {
+            flash.optsr_prg().modify(|_, w| w.swap_bank_opt().set_bit());
+        }
+
+        flash.optcr().modify(|_, w| w.optstart().set_bit());
+
+        loop {
+            if !flash.optsr_cur().read().opt_busy().bit() {
+                break;
+            }
+        }
+
+        ringbuf_entry!(Trace::FinishEnd);
+        Ok(())
+    }
+
+    fn poll_flash_done(&mut self) -> Result<(), RequestError<UpdateError>> {
+        let mut cnt = 0;
+        let mut seen = false;
+
+        loop {
+            if self.flash.bank2().sr.read().qw().bit() {
+                seen = true;
+                cnt = 0;
+            }
+
+            if seen && !self.flash.bank2().sr.read().qw().bit() {
+                break;
+            } else {
+                cnt += 1;
+            }
+
+            if cnt == FLASH_TIMEOUT {
```

Comment:
So, does the hardware specify a maximum programming time? Could we express this as a deadline (in time) that we will retry until? I worry that a canned counter like this could be fragile.

Reply:
Infinite delays seem ok if the hardware is genuinely hosed, we can always restart the task or reboot to attempt to recover.

This may prove to be famous last words.

Reply:
I removed the timeout.